src/pipeline.rs

```diff
@@ -1,18 +1,54 @@
 use std::fs::File;
 use std::io::{BufRead, BufReader};
 
-use crate::configuration::Configuration;
+use crate::configuration::{Column, Configuration, FileType};
 use crate::embedding::{calculate_embeddings, calculate_embeddings_mmap};
 use crate::entity::{EntityProcessor, SMALL_VECTOR_SIZE};
 use crate::persistence::embedding::TextFileVectorPersistor;
 use crate::persistence::entity::InMemoryEntityMappingPersistor;
 use crate::persistence::sparse_matrix::InMemorySparseMatrixPersistor;
 use crate::sparse_matrix::{create_sparse_matrices, SparseMatrix};
 use bus::Bus;
-use smallvec::SmallVec;
+use simdjson_rust::dom;
+use smallvec::{smallvec, SmallVec};
 use std::sync::Arc;
 use std::thread;
 
+/// Parse a line of JSON and read its columns into a vector for processing.
+fn read_json_columns(
+    line: &str,
+    parser: &mut dom::Parser,
+    columns: &[Column],
+) -> Vec<SmallVec<[String; SMALL_VECTOR_SIZE]>> {
+    let parsed = parser.parse(&line).unwrap();
+    columns
+        .iter()
+        .map({
+            |c| {
+                if !c.complex {
+                    smallvec![parsed
```

Comment:
We can use other types for non-complex column as well (something similar as it is for arrays):
```
let elem = parsed.at_key(&c.name).unwrap();
let value = match elem.get_type() {
  dom::element::ElementType::String => elem.get_string().unwrap(),
  _ => elem.minify(),
};
smallvec![value]
```

After modification we can read such lines:
```
{"users": 1, "products": ["p1", "p2"], "brands": ["b1", "b2"]}
{"users": 2, "products": ["p2", "p3", "p4"], "brands": ["b1"]}
```

Reply:
Okay. Although personally I would lean towards disallowing array values here since they should probably be treated as complex. This is why I only allowed strings initially. Although certainly other atomic values would make sense here.

Reply:
You are right but column config is here for "support". If there is some array it is treated as single entity and a. user see it in the output file. Otherwise the user should prepare json file with just strings.